graphql_compiler/compiler/ir_lowering_match.py

```diff
@@ -578,6 +593,194 @@ def prune_output_blocks_in_compound_match_query(compound_match_query):
+        if match_step.as_block.location in location_to_predicates:
+            where_block = Filter(
+                _predicate_list_to_where_block(
+                    location_to_predicates[match_step.as_block.location]
+                )
+            )
+            # Delete the location entry. No further filters needed for this location.
+            del location_to_predicates[match_step.as_block.location]
+        else:
+            where_block = None
+        new_match_step = MatchStep(
+            root_block=match_step.root_block,
+            coerce_type_block=match_step.coerce_type_block,
+            where_block=where_block,
+            as_block=match_step.as_block
+        )
+        new_match_traversal.append(new_match_step)
+    return new_match_traversal
+
+
+def collect_filters_to_first_location_instance(compound_match_query):
+    """Collate all filters for a particular location to the first instance of the location."""
+    new_match_queries = []
+    # Each MatchQuery is processed independently
+    for match_query in compound_match_query.match_queries:
+        location_to_predicates = _construct_location_to_predicate_list(match_query)
+
+        new_match_traversals = []
+        for match_traversal in match_query.match_traversals:
+            new_match_traversal = _collect_filters_to_first_location_in_match_traversal(
+                match_traversal, location_to_predicates)
+            new_match_traversals.append(new_match_traversal)
+        new_match_queries.append(
+            MatchQuery(
+                match_traversals=new_match_traversals,
+                folds=match_query.folds,
+                output_block=match_query.output_block
+            )
+        )
+
+    return CompoundMatchQuery(match_queries=new_match_queries)
+
+
+def _update_tagged_expression(expression, current_locations):
+    """Replace non-existent tag Expressons with True, and simplify the result."""
+    if isinstance(expression, BinaryComposition):
+        if isinstance(expression.left, ContextField):
+            context_field = expression.left
+        elif isinstance(expression.right, ContextField):
+            context_field = expression.right
+        elif expression.operator == u'||':
+            if expression.left == TrueLiteral or expression.right == TrueLiteral:
+                return TrueLiteral
+            else:
+                return expression
+        elif expression.operator == u'&&':
+            if expression.left == TrueLiteral and expression.right == TrueLiteral:
+                return TrueLiteral
+            if expression.left == TrueLiteral:
+                return expression.right
+            if expression.right == TrueLiteral:
+                return expression.left
+            else:
+                return expression
+        else:
+            return expression
+        location_name, _ = context_field.location.get_location_name()
```

Comment:
This is quite confusing. There are a ton of `return` statements above this, including one in an `else` clause right above, and yet here we have more code. It's confusing and far from obvious that two of the if-elif cases actually fall through to here.

A few suggestions:
- Break up the `if/elif/else` cascade into clearly disjoint cases, then cut off the biggest subtrees and make them helper functions.
- Make helpers for "BinaryComposition with a ContextField as the left/right expression", "BinaryComposition with &&", "BinaryComposition with ||", "TernaryConditional".
- Add checks and AssertionErrors for expressions that you don't expect to see, and therefore don't handle. This includes not just `BetweenClause` but also things like `OutputContextField`, `FoldedOutputContextField` etc.